test/distributed/fsdp/test_fsdp_optim_state.py

```diff
@@ -433,39 +486,56 @@ def _test_shard_full_optim_state(
             else self._init_transformer_model if model_class == "transformer" \
             else None
         assert initializer is not None, f"Unsupported model: {model_class}"
-        # Run a wrapped model with full world size for a few iterations
+        # First, run a wrapped model with full world size for a few iterations
         model1, optim1, optim_input1 = initializer(
             wrap=True, use_multiple_param_groups=use_multiple_param_groups,
         )
         self._step_model(model1, optim1, num_iters=NUM_ITERS)
         full_osd1 = FSDP.full_optim_state_dict(model1, optim1, optim_input1)
-        # Broadcast instead of `torch.save()`/`torch.load()` so that all ranks
-        # have the full state dict
-        full_osd1 = self._broadcast_full_osd(full_osd1)
         if halve_world_size:
             # Create a new process group with halved world size
             new_group_ranks = [r for r in range(self.world_size) if r % 2 == 0]
             new_group = dist.new_group(ranks=new_group_ranks)
             if self.rank not in new_group_ranks:
                 return
         else:
+            # Continue using the same group and hence world size
             new_group = dist.distributed_c10d._get_default_group()
-        # Run a wrapped model with halved world size (from scratch)
+        # Second, run a wrapped model with (possibly) halved world size
         model2, optim2, optim_input2 = initializer(
             wrap=True, group=new_group,
             use_multiple_param_groups=use_multiple_param_groups,
             **new_model_kwargs,  # specify `wrap_alt` to change wrapping
         )
         self._step_model(model2, optim2, num_iters=NUM_ITERS)
         full_osd2 = FSDP.full_optim_state_dict(model2, optim2, optim_input2)
-        full_osd2 = self._broadcast_full_osd(full_osd2, group=new_group)
-        # As a sanity check, check that sharding the halved-world-size model's
-        # full optimizer state dict according to itself is equivalent to its
-        # local optimizer's state dict
+        # Compute two sharded optim state dicts: (1) for the first model
+        # according to the second model and (2) for the second model according
+        # to the second model
```

Comment:
We see below how calling `FSDP.scatter_full_optim_state_dict()` is equivalent to broadcasting the full optimizer state dict and calling `FSDP.shard_full_optim_state_dict()`.

In other words, `sharded_osd1` and `sharded_osd2` are the same regardless of `osd_comm_method` (except for possibly the tensors' residing device).